src/Microsoft.AspNet.SignalR.ServiceBus3/ServiceBusConnection.cs

```diff
@@ -0,0 +1,326 @@
+                TopicClient topicClient = TopicClient.CreateFromConnectionString(_connectionString, topicName);
+
+                if (_configuration.RetryPolicy != null)
+                {
+                    topicClient.RetryPolicy = _configuration.RetryPolicy;
+                }
+                else
+                {
+                    topicClient.RetryPolicy = RetryExponential.Default;
+                }
+
+                connectionContext.SetTopicClients(topicClient, topicIndex);
+
+                _trace.TraceInformation("Creation of a new topic client {0} completed successfully.", topicName);
+            }
+
+            CreateSubscription(connectionContext, topicIndex);
+        }
+
+        private void CreateSubscription(ServiceBusConnectionContext connectionContext, int topicIndex)
+        {
+            lock (connectionContext.SubscriptionsLock)
+            {
+                if (connectionContext.IsDisposed)
+                {
+                    return;
+                }
+
+                string topicName = connectionContext.TopicNames[topicIndex];
+
+                // Create a random subscription
+                string subscriptionName = Guid.NewGuid().ToString();
+
+                try
+                {
+                    var subscriptionDescription = new SubscriptionDescription(topicName, subscriptionName);
+
+                    // This cleans up the subscription while if it's been idle for more than the timeout.
+                    subscriptionDescription.AutoDeleteOnIdle = _idleSubscriptionTimeout;
+
+                    _namespaceManager.CreateSubscription(subscriptionDescription);
+
+                    _trace.TraceInformation("Creation of a new subscription {0} for topic {1} in the service bus completed successfully.", subscriptionName, topicName);
+                }
+                catch (MessagingEntityAlreadyExistsException)
+                {
+                    // The entity already exists
+                    _trace.TraceInformation("Creation of a new subscription {0} for topic {1} threw an MessagingEntityAlreadyExistsException.", subscriptionName, topicName);
+                }
+
+                // Create a receiver to get messages
+                string subscriptionEntityPath = SubscriptionClient.FormatSubscriptionPath(topicName, subscriptionName);
+                MessageReceiver receiver = _factory.CreateMessageReceiver(subscriptionEntityPath, ReceiveMode.ReceiveAndDelete);
+
+                _trace.TraceInformation("Creation of a message receive for subscription entity path {0} in the service bus completed successfully.", subscriptionEntityPath);
+
+                connectionContext.SetSubscriptionContext(new SubscriptionContext(topicName, subscriptionName, receiver), topicIndex);
+
+                var receiverContext = new ReceiverContext(topicIndex, receiver, connectionContext);
+
+                Task.Run(() => ProcessMessages(receiverContext));
```

Comment:
We didn't before so I did not look into it. I am not sure what would be the consequences of synchronously blocking here.  I guess it would also have to have time out in which case if the timeout expires we are in the same boat - Dispose() finished task is not finished. But I will look into it.